File: 73-set-matrix-zeroes/set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& A) {
        int m = A.size(), n = A[0].size();
        bool zerothRow = false;
        for(int i=0;i<n;i++){
            if(A[0][i]==0){
                zerothRow = true;
                break;
            }
        }
        for(int i=1;i<m;i++){
            for(int j=0;j<n;j++){
                if(A[i][j]==0){
                    A[i][0]=0;
                    A[0][j]=0;
                }
            }
        }
        for(int i=m-1;i>0;i--){
            for(int j=n-1;j>=0;j--){
                if(A[i][0]==0 || A[0][j]==0) A[i][j]=0;
            }
        }
        if(zerothRow){
            for(int i=0;i<n;i++) A[0][i]=0;
        }
    }
};

File: 73-set-matrix-zeroes/set-matrix-zeroes.java
class Solution {
    public void setZeroes(int[][] matrix) {
        int numRow= matrix.length, numCol= matrix[0].length;
        boolean isFirstRowZero= false;
        for(int j=0;j<numCol;j++){
            if(matrix[0][j]==0){
                isFirstRowZero= true;
                break; // no need to iterate further
            }
        }
        for(int i=1;i<numRow;i++){
            for(int j=0;j<numCol;j++){
                if(matrix[i][j]==0){
                    matrix[i][0]= 0;
                    matrix[0][j]= 0;
                }
            }
        }
        for(int i=numRow-1;i>0;i--){
            for(int j=numCol-1;j>=0;j--){
                if(matrix[i][j]==0) continue;
                if(matrix[i][0]==0 || matrix[0][j]==0){
                    matrix[i][j]=0;
                }
            }
        }
        if(isFirstRowZero){
            for(int j=0;j<numCol;j++){
                matrix[0][j]= 0;
            }
        }
    }
}